Refactoring of Records object initialization

mmessick opened the pull request:
More flexible initialization that will not require any data file. This will allow easier integration with Luca when we get there. Thus, we can just create a blank Records object and take in values from a filled out .json form, setting these values to their respective ecodes accordingly

mmessick commented:
not passing the Travis CI tests, but passes all tests on my computer

mmessick commented:
@talumbau, if you have time, could you take a look at the diffs? I extracted the methods read_weights and read_blowup so that they are their own functions in records.py. I was able to change the paths for various tests so that this worked smoothly on my computer, however, the Travis CI build doesn't seem to be working despite this fact. If you have any advice or suggestions on this issue, that would be great. Thanks!

talumbau commented:
@mmessick it looks like one issue is the location of some of the test files (e.g. `tax_all1991_puf.gz`, `WEIGHTS_testing.csv`, etc.). The way this is done in the tests is as follows:

`CUR_PATH` is set to be the directory containing the given test file:

```
CUR_PATH = os.path.abspath(os.path.dirname(__file__))
```

So, let's call that `./taxcalc/tests/`. On my computer, in the repo that I use frequently, that directory is something like `/Users/talumbau/dev/aei/Tax-Calculator/taxcalc/tests`. The test files are two directories "up" from that path, so that's why you see code like this:

```
WEIGHTS_FILENAME = "../../WEIGHTS_testing.csv"
weights_path = os.path.join(CUR_PATH, WEIGHTS_FILENAME)
```

So, on my computer, `weights_path` would then be: `/Users/talumbau/dev/aei/Tax-Calculator/WEIGHTS_testing.csv`, which is indeed where the file is. Notice that this is independent of where one runs the tests because it is based on the location of the test file (e.g. `test_calculator.py`) and not the current directory when you invoke pytest. That being said, the best way to run all of the tests is to be at the top of the repo (e.g. `/Users/talumbau/dev/aei/Tax-Calculator/`) and then just type `py.test`. If that results in errors on your machine, then Travis CI will not work because it does exactly the same thing. One thing to be sure of is that you do not have the `taxcalc` package installed in the environment where you are testing. It will use the installed package instead of the code you are working on for the tests. You can check if `taxcalc` is installed by typing `conda list`. If you are at the top of the repo and you want to make sure that the `taxcalc` for your tests is the current code in your repo (and not some earlier version), after typing `conda list | grep 'taxcalc'`, you should see something like:

```
taxcalc (/Users/talumbau/dev/aei/Tax-Calculator) 0.4.9 5 g0a9f808 dirty           <pip>
```
indicating that the taxcalc package that was found in the local repo, not an installed package.

So I think the first step in debugging this is to get the tests passing in the way I describe above on your machine. Let me know how that goes.

talumbau commented on taxcalc/records.py:
```diff
@@ -25,31 +32,178 @@ class Records(object):
+        if data is not None:
+            self.read(data)
+        else:
+            assert type(dims) is not None
```
I'm not sure this is what you want here. Even if `dims` is `None` (which is the default value), the type of `dims` is a special type, `NoneType` which is not `None` so this assertion will pass. In fact, I can't think of any case where this assertion would fail (since even `None` has a type), so I don't think the assertion checks for any useful case. What is the desired check here?

mmessick replied:
You're right. This should be 

    assert dims is not None

talumbau replied:
In that case, I think you want:

```
assert dims is not None
self.read(data)
```
Since you will kill the program if dims is None, might as well check for that first, and then proceed as if it is not None. However, this is a significant change to the creation of a `Records` object. What is `dims` and why is it necessary for every `Records` object creation?

talumbau replied:
Oops. disregard above. I misread the code.

talumbau replied:
You are saying that it can not be the case that both `data` and `dims` are `None`, right? Then I would suggest:

```
if data is None and dims is None:
    #quit program or raise excpetion here
    raise ValueError("some message)
    #or just assert False
    assert False

mmessick replied:
Yes. Ok I will add this--thanks!

talumbau commented on taxcalc/records.py:
```diff
@@ -25,31 +32,178 @@ class Records(object):
+        self.BF = read_blowup(blowup_factors)
+        self.WT = read_weights(weights)
+
+        attr = ['AGIR1', 'DSI', 'EFI', 'EIC', 'ELECT', 'e04600',
```
We should only have this list in here once. Actually I think the formatting of this list is better than the one below, but we need to find a way to not list all of these values twice.

talumbau replied:
In particular, the list here:
https://github.com/mmessick/Tax-Calculator/blob/luca/taxcalc/records.py#L467
is the same list

talumbau commented on taxcalc/records.py:
```diff
@@ -25,31 +32,178 @@ class Records(object):
+                'e87510', 'e87520', 'e87530', 'e87540', 'e87550',
+                'RECID', 's006', 's008', 's009', 'WSAMP', 'TXRT']
+
+        zero = ['e35300_0', 'e35600_0', 'e35910_0', 'x03150', 'e03600',
```
same comment as above. The list here:
https://github.com/mmessick/Tax-Calculator/blob/luca/taxcalc/records.py#L682
is the same as this list.

talumbau commented on taxcalc/records.py:
```diff
@@ -659,10 +785,60 @@ def read(self, data):
+        # and untaxed voluntary contrib's to retirement plans
+        # (e07240, not sure if from employer or voluntary)
+
+    # TODO: not sure if we have good data on insurance or retirment
```
these comments should be lined up with the indentation level for this function

talumbau commented on taxcalc/records.py:
```diff
@@ -659,10 +785,60 @@ def read(self, data):
+        # individual's share of the dividends received
+        self.dividends = self.e00600
+
+    # JTC thinks that dividends are enough to measure stock ownership,
```
these comments should be lined up with the indentation level for this function

talumbau commented on taxcalc/records.py:
```diff
@@ -684,3 +860,64 @@ def imputation(e17500, e00100, e18400, e18425, e62100, e00700, e04470,
+            print("Missing a csv file with blowup factors. Please pass "
+                  "such a csv as PUF(blowup_factors='[FILENAME]').")
+            raise
+            # TODO, we will need to pass the csv to the Calculator once
```
this comment is old and should have been removed a while ago. Best to get rid of it now.

talumbau commented on taxcalc/records.py:
```diff
@@ -684,3 +860,64 @@ def imputation(e17500, e00100, e18400, e18425, e62100, e00700, e04470,
+                  "data of the data extrapolation. Please pass such a "
+                  "file as PUF(weights='[FILENAME]').")
+            raise
+            # TODO, we will need to pass the csv to the Calculator once
```
see above about old comment.

mmessick commented:
@talumbau, seems to be working now. I discovered that the error for the Travis CI build had to do with where I was declaring the path names. Since I moved those declarations to outside the Records object, they were always associated with the folder that records.py was in, not where the Records object got created. I fixed this by moving those back to Records object attributes

talumbau commented:
Ok, looks like the tests are passing so that's good. In the latest code though, I don't see any tests that show how one would use this new capability. Can you add some tests to `taxcalc/tests/test_records.py` that show the impact of your changes? This is even better than documentation since it shows someone exactly how to use the new capabilities.

mmessick commented:
The basic idea when using this in concurrence with a Luca JSON file or with @martinholmer's integration of having an individual taxpayer input his/her own tax file would to be to initialize a zeroed out Records object. Then a functions would be easy to write within our calculator that would work with either way of inputting individual tax records. We would just set the record's variables to corresponding values on the forms.

Example for an individual taxpayer:

    r = Records(dims=1)
    r.set_attr("e00100", 1000)  # sets an individual value
    for name, data in form:  # this would be a loop that could go through the values for an indvidual's taxes
        r.set_attr(name, data)

feenberg commented:
On Fri, 17 Jul 2015, mmessick wrote:

> 
> The basic idea when using this in concurrence with a Luca JSON file or with
> @martinholmer's integration of having an individual taxpayer input his/her own tax
> file would to be to initialize a zeroed out Records object. Then a functions would be
> easy to write within our calculator that would work with either way of inputting
> individual tax records. We would just set the record's variables to corresponding
> values on the forms.
> 
> Example for an individual taxpayer:
> 
> r = Records(dims=1)
> r.set_attr("e00100", 1000)  # sets an individual value
> for name, data in form:  # this would be a loop that could go through the values for a
> n indvidual's taxes
>     r.set_attr(name, data)

I am confused. Why do we need e00100? We calculate it, Is the data value 
used for the tab variable? Is this function about allowing a user to do 
the tax calculation for a single taxpayer, or about how user supplied data 
will be transferred from CPS or other files? That should be a discussion 
item. I found Luca on Google - what will it be doing for us? Would anyone 
like to fill me in on what is going on?

dan
617-863-0343


> 
> —
> Reply to this email directly or view it on
> GitHub.[AHvQVcOPLmKGL3nL5wEi9e819a1H4eKFks5oeT9lgaJpZM4FXpCq.gif]
> 
> 
>

MattHJensen commented:
@feenberg, you're right, there is not a good explanation of the purpose behind this project. I just opened issue #319 to rectify that. 

You are also right, of course, that we calculate `e00100` and do not need to input it. `e00200` would go better in this example

martinholmer commented:
@mmessick, I am finally far enough along in my work related to issue #291 that I would like to try out your pull request #309, which looks promising for my needs.  However, I am sorry to say that I've taken so long getting to this stage of my work, that your pull request #309 is behind the master branch (mostly with respect to the PEP8 compliance work done recently by @MattHJensen).  I'm hoping you can bring your luca branch up-to-date so that I don't get merge conflicts when I pull your luca branch into my repo, which is up-to-date with the OSPC master branch.  If you don't have time to do this now, please let me know so that I can continue my work using a different Records class strategy.  Thanks in advance for any   help you can provide now.

mmessick commented:
@martinholmer, I am working on merging it now. I will probably have it ready by either tomorrow or Monday

mmessick commented:
@martinholmer, I am running to a little more issues with merging than I expected and I have a lot of assignments due tomorrow, but I will complete it tomorrow

martinholmer commented:
@mmessick, Thanks for your effort to bring this pull request up-to-date with the master.  When you finish I will be able to evaluate it in more detail.

Meanwhile, I thought I should share some very preliminary findings from my work on issue #291.  These findings could well be wrong, but if they are correct, they suggest we can achieve our goal of working with the OSPC Tax-Calculator using a single-filing-unit in any year beginning with 2013 (the first year for which there are tax policy parameters) more easily than we realized before.  There is one key part of your pull request #309 that I got into the master branch today (as closed pull request #352): the part that moves the NAMES list out of the read() method into Records class scope where it is publicly available.  That was absolutely essential as anyone reading your pull request should have realized.

Below is a short script that creates a single-filing-unit Records object and uses it to create a Calculator object.  Then the script specifies a simple 2014 filing unit, calculates its tax liability, and compares it to that unit's tax liability as computed by Internet TAXSIM.  Then the script specifies a simple 2013 filing unit, calculates its tax liability, and compares it to that unit's tax liability as computed by Internet TAXSIM.  Both filing-unit's tax liabilities are the same down to the penny.

I'm not happy about the pylint warnings that tell me the script is assessing a private member (that is, the _numextra and _ospctax variables), but the Python interpreter ignores this violation of privacy and computes the correct tax liability anyway.  I figure we can fix these kinds of things later after we have more confidence that we are doing this single-filing-unit work correctly.

I'm interested in your candid reactions to the approach this script takes.  And I'm also interested, when you get the time, in hearing about whether you can use this same kind of approach in your Luca work.

The script is running under Python 2.7 on my computer, so if you're on 3.4 you'll need to fix my old-fashioned print statements.  Also, if you put the script in an empty directory parallel with your tax-calculator directory, you will not need to change the path to the Tax-Calculator classes.   And, of course, you need the master branch from late Monday afternoon in order to get the publicly-available NAMES list.

Here is the script:
```
"""
Test using Tax-Calculator package with one-filing-unit Records object.

NOTE: blowup can be avoided by not calling Records.increment_year() method
"""
# CODING-STYLE CHECKS:
# pep8 --ignore=E402 test1.py
# pylint --disable=locally-disabled test1.py

import os
import sys
from numpy.core import int32 as np_int32
import pandas as pd
CURRENT_PATH = os.path.abspath(os.path.dirname(__file__))
sys.path.append(os.path.join(CURRENT_PATH, "../tax-calculator"))
# pylint: disable=import-error
from taxcalc import Parameters, Records, Calculator


def main():
    """
    Complete logic of the script.
    """
    # construct Calculator object for 2013 with a single tax filing unit
    prm = Parameters()
    assert prm.current_year == 2013
    dfd = {}
    for names in Records.NAMES:
        dfd[names[1]] = [0]
    dfd['flpdyr'] = [2013]  # must be equal to parameters current_year
    recdata = pd.DataFrame.from_dict(dfd, dtype=np_int32)
    rec = Records(data=recdata)
    assert rec.current_year == 2013
    assert rec.dim == 1
    calc = Calculator(params=prm, records=rec)
    assert calc.current_year == 2013

    # pylint: disable=protected-access

    # specify ID=21 (see below)
    calc.records.FLPDYR[0] = 2014
    calc.records.MARS[0] = 1
    calc.records.XTOT[0] = 1
    calc.records._numextra[0] = 0
    calc.records.e00200[0] = 18000
    print 'ID=21 ', \
        ' YEAR=', calc.records.FLPDYR[0], \
        ' MARS=', calc.records.MARS[0], \
        ' XTOT=', calc.records.XTOT[0], \
        ' numextra=', calc.records._numextra[0], \
        ' FDED=', calc.records.FDED[0], \
        ' WAGE=', calc.records.e00200[0]
    calc.params.set_year(calc.records.FLPDYR[0])
    print 'calc.params.current_year=', calc.params.current_year
    calc.calc_all()
    expected_tax = 785.00  # as for ID=21 TAXSIM ovar[4]
    calculated_tax = calc.records._ospctax[0]
    if calculated_tax == expected_tax:
        print '*** OK(ID=21) ***'
    else:
        print '### ERROR(ID=21): ', \
            'SIMTAX=', calculated_tax, \
            'TAXSIM=', expected_tax

    # specify ID=11 (see below)
    calc.records.FLPDYR[0] = 2013
    calc.records.MARS[0] = 1
    calc.records.XTOT[0] = 1
    calc.records._numextra[0] = 0
    calc.records.e00200[0] = 58000
    print 'ID=11 ', \
        ' YEAR=', calc.records.FLPDYR[0], \
        ' MARS=', calc.records.MARS[0], \
        ' XTOT=', calc.records.XTOT[0], \
        ' numextra=', calc.records._numextra[0], \
        ' FDED=', calc.records.FDED[0], \
        ' WAGE=', calc.records.e00200[0]
    calc.params.set_year(calc.records.FLPDYR[0])
    print 'calc.params.current_year=', calc.params.current_year
    calc.calc_all()
    expected_tax = 7928.75  # as for ID=11 TAXSIM ovar[4] (see below)
    calculated_tax = calc.records._ospctax[0]
    if calculated_tax == expected_tax:
        print '*** OK(ID=11) ***'
    else:
        print '### ERROR(ID=11): ', \
            'SIMTAX=', calculated_tax, \
            'TAXSIM=', expected_tax

    # return no-error exit code
    return 0
# end of main function code


if __name__ == '__main__':
    sys.exit(main())


# TAXSIM INPUT:
# 11 2013 0 1 0 0 58000 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
# 37 2015 0 3 1 1 46000 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
# 21 2014 0 1 0 0 18000 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
# 19 2013 0 2 0 1 49000 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
# 27 2015 0 3 2 0 32000 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
# TAXSIM OUTPUT:
# 11. 2013 0 7928.75 .00 8874.00 25.00 .00 15.30 58000.00
# .00 .00 6100.00 3900.00 .00 .00 .00 48000.00 7928.75 .00 .00
# .00 .00 .00 .00 58000.00 .00 7928.75 8874.00
# 37. 2015 0 3422.50 .00 7038.00 15.00 .00 15.30 46000.00
# .00 .00 10800.00 8000.00 .00 .00 .00 27200.00 3422.50 .00 .00
# .00 .00 .00 .00 46000.00 .00 3422.50 7038.00
# 21. 2014 0 785.00 .00 2754.00 10.00 .00 15.30 18000.00
# .00 .00 6200.00 3950.00 .00 .00 .00 7850.00 785.00 .00 .00
# .00 .00 .00 .00 18000.00 .00 785.00 2754.00
# 19. 2013 0 3277.50 .00 7497.00 15.00 .00 15.30 49000.00
# .00 .00 13400.00 7800.00 .00 .00 .00 27800.00 3277.50 .00 .00
# .00 .00 .00 .00 49000.00 .00 3277.50 7497.00
# 27. 2015 0 -1488.80 .00 4896.00 31.06 .00 15.30 32000.00
# .00 .00 9250.00 12000.00 .00 .00 .00 10750.00 1075.00 .00 .00
# .00 .00 .00 2563.80 32000.00 .00 1075.00 4896.00


# RESULTS OF THIS SCRIPT:
# iMac2:simtax mrh$ time python test1.py
# ID=21   YEAR= 2014  MARS= 1  XTOT= 1  numextra= 0.0  FDED= 0  WAGE= 18000
# calc.params.current_year= 2014
# *** OK(ID=21) ***
# ID=11   YEAR= 2013  MARS= 1  XTOT= 1  numextra= 0.0  FDED= 0  WAGE= 58000
# calc.params.current_year= 2013
# *** OK(ID=11) ***
#
# real	0m29.252s
# user	0m28.897s
# sys	0m0.341s
```

mmessick commented:
Local tests pass, just not PEP 8, will fix those quickly in the morning but then we should be good to go with this PR.

@martinholmer, when I get more time later in the week, I will check out your script. Looks promising. Although I won't have a use for it since I have finished my summer internship and am no longer working on Luca. However, I'm sure it will be helpful for whoever picks up that project next

martinholmer commented:
@mmessick,  Thanks again for all the work.  I didn't know your tenure on this project was limited to a summer intership.  There is no need on your end to pursue this any further; the pull request was very helpful in stimulating my thinking on the TAXSIM-like input/output project.

martinholmer commented:
@MattHJensen,
You might consider closing this pull request.  A key element of this pull request #309 was merged into the master branch as part of pull request #352.  Other than that essential change, I have found that the current version of the Records class can be used in one-filing-unit situations as with Luca and with the new TAXSIM-like capability described in issue #291.  All the other changes being proposed in this pull request are not necessary for my work on the TAXSIM-like capability.

MattHJensen commented:
@martinholmer thanks. And thank you @mmessick!
